qiskit/algorithms/optimizers/steppable_optimizer.py

```diff
@@ -0,0 +1,213 @@
+        Canonical optimization workflow using ask() and tell() can be seen in SteppableOptimizer.step().
+        Args:
+            ask_object: Contains the information on how the evaluation was done.
+            tell_object: Contains all relevant information about the evaluation of the objective function.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def evaluate(self, ask_object: AskObject) -> TellObject:
+        """
+        Evaluates the function according to the instructions contained in ask_object.
+        If the user decides to use step() instead of ask() and tell() this function will contain the logic on how to evaluate
+        the function.
+        Args:
+            ask_object: Contains the information on how to do the evaluation.
+        Return:
+            tell_object: Contains all relevant information about the evaluation of the objective function.
+        """
+        raise NotImplementedError
+
+    def user_evaluate(
+        self, *args, **kwargs
+    ) -> TellObject:  # Maybe this function should be abstract. For some optimizers (The return type with a custom tell object the function needs to be overwriten)
+        """
+        Constructs TellObject.
+        Used when the user manually evaluates the function.
+        """
+        return TellObject(*args, **kwargs)
+
+    def step(self) -> None:
+        """
+        Performs one step in the optimization process.
+        This mehtod composes ask(), evaluate(), and tell() to make a step in the optimization process.
+        """
+        ask_object = self.ask()
+        tell_object = self.evaluate(ask_object=ask_object)
+        self.tell(ask_object=ask_object, tell_object=tell_object)
+        if self.callback is not None:
+            self.callback(state=self._state)
+
+    def minimize(
+        self,
+        fun: Callable[[POINT], float],
+        x0: POINT,
+        jac: Optional[Callable[[POINT], POINT]] = None,
+        bounds: Optional[List[Tuple[float, float]]] = None,
+        **kwargs,
+    ) -> OptimizerResult:
+        """
+        For well behaved functions the user can call this method to minimize a function. If the user wants more control
+        on how to evaluate the function a custom loop can be created using ask() and tell() and evaluating the function
+        manually.
+        """
+        self.initialize(x0=x0, fun=fun, jac=jac)
+        for _ in range(self.maxiter):
+            self.step()
+            if self.stop_condition():
+                break
+
+        return self.create_result()
+
+    @abstractmethod
+    def create_result(self) -> OptimizerResult:
+        """
+        Returns the result of the optimization.
+        All the information needed to create the result should be stored in the optimizer state.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def initialize(self, *args, **kwargs) -> None:
```

Comment:
I think this should have the same signature as `minimize`, so that that the following two usages are the same:
```python
result = opt.minimize(f, x0)
```
and 
```python
opt.initialize(f, x0)
while not opt.stop_criterion():
     opt.step()
```
All the hyperparameters are currently stored in the `Optimizer.__init__` and to not restrict hyperparameter setting to `initialize` we should keep it this way 🙂